File: app/src/main/java/com/dougkeen/bart/model/Route.java
package com.dougkeen.bart.model;

import java.util.Collection;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Route {
    private Station origin;
    private Station destination;
    private Line directLine;
    private Collection<Line> transferLines;
    private boolean requiresTransfer;
    private Station transferStation;
    private String direction;
    private List<Line> lines = new ArrayList<Line>();
    private List<Station> transferStations = new ArrayList<Station>();
    private final Map<Line, List<Station>> stationSequencesByLine =
            new HashMap<Line, List<Station>>();

    public Station getOrigin() {
        return origin;
    }

    public void setOrigin(Station origin) {
        this.origin = origin;
    }

    public Station getDestination() {
        return destination;
    }

    public void setDestination(Station destination) {
        this.destination = destination;
    }

    public Line getDirectLine() {
        return directLine;
    }

    public void setDirectLine(Line line) {
        this.directLine = line;
        if (lines.isEmpty() && line != null) {
            lines.add(line);
        }
    }

    public Collection<Line> getTransferLines() {
        return transferLines;
    }

    public void setTransferLines(Collection<Line> transferLines) {
        this.transferLines = transferLines;
        if (lines.size() <= 1 && transferLines != null) {
            for (Line line : transferLines) {
                if (!lines.contains(line)) {
                    lines.add(line);
                }
            }
        }
    }

    public boolean hasTransfer() {
        return requiresTransfer;
    }

    public void setTransfer(boolean requiresTransfer) {
        this.requiresTransfer = requiresTransfer;
    }

    public Station getTransferStation() {
        return transferStation;
    }

    public void setTransferStation(Station transferStation) {
        this.transferStation = transferStation;
        if (transferStation != null && transferStations.isEmpty()) {
            transferStations.add(transferStation);
        }
    }

    public String getDirection() {
        return direction;
    }

    public void setDirection(String direction) {
        this.direction = direction;
    }

    public List<Line> getLines() {
        return Collections.unmodifiableList(lines);
    }

    public void setLines(List<Line> lines) {
        this.lines = new ArrayList<Line>();
        if (lines != null) {
            this.lines.addAll(lines);
        }
        if (directLine == null && !this.lines.isEmpty()) {
            directLine = this.lines.get(0);
        }
    }

    public List<Station> getTransferStations() {
        return Collections.unmodifiableList(transferStations);
    }

    public void setTransferStations(List<Station> transferStations) {
        this.transferStations = new ArrayList<Station>();
        if (transferStations != null) {
            this.transferStations.addAll(transferStations);
        }
        transferStation = this.transferStations.isEmpty()
                ? null : this.transferStations.get(0);
    }

    /** Associates a route with the feed pattern used to build it. */
    public void setStationSequence(Line line, List<Station> stations) {
        if (line == null || stations == null || stations.isEmpty()) {
            return;
        }
        stationSequencesByLine.put(line,
                Collections.unmodifiableList(new ArrayList<Station>(stations)));
    }

    /** Returns the feed pattern used for this line. */
    public List<Station> getStationSequence(Line line) {
        return stationsForLine(line);
    }

    private List<Station> stationsForLine(Line line) {
        List<Station> sequence = stationSequencesByLine.get(line);
        return sequence == null ? Collections.<Station>emptyList() : sequence;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Route [origin=");
        builder.append(origin);
        builder.append(", destination=");
        builder.append(destination);
        builder.append(", line=");
        builder.append(directLine);
        builder.append(", requiresTransfer=");
        builder.append(requiresTransfer);
        builder.append(", transferStation=");
        builder.append(transferStation);
        builder.append(", direction=");
        builder.append(direction);
        builder.append("]");
        return builder.toString();
    }

    public boolean trainDestinationIsApplicable(Station lineDestination,
                                                Line viaLine) {
        List<Station> viaStations = stationsForLine(viaLine);
        int originIndex = viaStations.indexOf(origin);
        if (destination == null) {
            return originIndex >= 0 && lineDestination != null
                    && viaLine.equals(directLine)
                    && viaStations.indexOf(lineDestination) >= 0
                    && lineDestination != origin;
        }
        int routeDestinationIndex = viaStations.indexOf(destination);
        int lineDestinationIndex = viaStations.indexOf(lineDestination);

        boolean hasDirectRouteViaLine = originIndex >= 0 && routeDestinationIndex >= 0;

        if (requiresTransfer && !lines.isEmpty()) {
            if (transferStations.isEmpty() && directLine != null
                    && directLine.requiresTransfer()) {
                return viaLine.equals(directLine.transferLine1)
                        || viaLine.equals(directLine.transferLine2);
            }
            // Departures are queried at the passenger origin, so only the
            // first line of a multi-leg route can be represented by an ETD
            // returned for this route.  The remaining lines are used later
            // when the connecting trips are paired with this departure.
            if (!viaLine.equals(lines.get(0))) {
                return false;
            }
            Station firstTransfer = transferStations.isEmpty()
                    ? destination : transferStations.get(0);
            int transferIndex = viaStations.indexOf(firstTransfer);
            if (originIndex < 0 || transferIndex < 0 || lineDestinationIndex < 0) {
                return false;
            }
            int direction = Integer.compare(transferIndex, originIndex);
            return direction != 0
                    && Integer.compare(lineDestinationIndex, originIndex)
                    == direction
                    && Integer.compare(lineDestinationIndex, transferIndex)
                    * direction >= 0;
        } else {
            return originIndex >= 0 && routeDestinationIndex >= 0
                    && lineDestinationIndex >= 0
                    && ((originIndex <= routeDestinationIndex
                    && routeDestinationIndex <= lineDestinationIndex
                    && lineDestinationIndex >= originIndex)
                    || (originIndex >= routeDestinationIndex
                    && routeDestinationIndex >= lineDestinationIndex
                    && lineDestinationIndex <= originIndex));
        }
    }

}

File: app/src/main/kotlin/in/izyum/bart/model/Route.kt
package `in`.izyum.bart.model

import java.util.ArrayList
import java.util.Collections
import java.util.LinkedHashMap

/** Immutable route topology and the GTFS patterns used to construct it. */
class Route private constructor(
    val origin: Station?,
    val destination: Station?,
    val directLine: Line?,
    transferLines: List<Line>,
    val requiresTransfer: Boolean,
    val direction: String?,
    lines: List<Line>,
    transferStations: List<Station>,
    stationSequencesByLine: Map<Line, List<Station>>
) {
    val transferLines: List<Line> = immutableRouteList(transferLines)
    val lines: List<Line> = immutableRouteList(lines)
    val transferStations: List<Station> = immutableRouteList(transferStations)
    private val stationSequences: Map<Line, List<Station>> = immutableRouteMap(
        stationSequencesByLine
    )

    val transferStation: Station?
        get() = transferStations.firstOrNull()

    fun hasTransfer(): Boolean = requiresTransfer

    /** Returns the feed pattern used for this line. */
    fun getStationSequence(line: Line?): List<Station> =
        stationSequences[line].orEmpty()

    /**
     * Returns whether a train terminating at [lineDestination] can serve this
     * route leg on [viaLine].
     */
    fun trainDestinationIsApplicable(
        lineDestination: Station?,
        viaLine: Line?
    ): Boolean {
        val viaStations = getStationSequence(viaLine)
        val originIndex = viaStations.indexOf(origin)
        if (destination == null) {
            return originIndex >= 0
                && lineDestination != null
                && viaLine == directLine
                && viaStations.indexOf(lineDestination) >= 0
                && lineDestination != origin
        }

        val routeDestinationIndex = viaStations.indexOf(destination)
        val lineDestinationIndex = viaStations.indexOf(lineDestination)
        val hasDirectRouteViaLine = originIndex >= 0 && routeDestinationIndex >= 0

        if (requiresTransfer && lines.isNotEmpty()) {
            if (transferStations.isEmpty() && directLine?.requiresTransfer() == true) {
                return viaLine == directLine.transferLine1
                    || viaLine == directLine.transferLine2
            }
            // Departures are queried at the passenger origin, so only the
            // first line of a multi-leg route can be represented by an ETD
            // returned for this route. The remaining lines are paired later.
            if (viaLine != lines.first()) {
                return false
            }
            val firstTransfer = transferStations.firstOrNull() ?: destination
            val transferIndex = viaStations.indexOf(firstTransfer)
            if (originIndex < 0 || transferIndex < 0 || lineDestinationIndex < 0) {
                return false
            }
            val direction = transferIndex.compareTo(originIndex)
            return direction != 0
                && lineDestinationIndex.compareTo(originIndex) == direction
                && lineDestinationIndex.compareTo(transferIndex) * direction >= 0
        }

        return hasDirectRouteViaLine
            && lineDestinationIndex >= 0
            && ((originIndex <= routeDestinationIndex
            && routeDestinationIndex <= lineDestinationIndex
            && lineDestinationIndex >= originIndex)
            || (originIndex >= routeDestinationIndex
            && routeDestinationIndex >= lineDestinationIndex
            && lineDestinationIndex <= originIndex))
    }

    override fun toString(): String =
        "Route [origin=$origin, destination=$destination, line=$directLine, " +
            "requiresTransfer=$requiresTransfer, transferStation=$transferStation, " +
            "direction=$direction]"

    companion object {
        @JvmStatic
        fun stationOnly(
            origin: Station,
            line: Line,
            stationSequence: List<Station>
        ): Route = Route(
            origin = origin,
            destination = null,
            directLine = line,
            transferLines = emptyList(),
            requiresTransfer = false,
            direction = null,
            lines = listOf(line),
            transferStations = emptyList(),
            stationSequencesByLine = mapOf(line to stationSequence)
        )

        @JvmStatic
        fun direct(
            origin: Station,
            destination: Station,
            line: Line,
            direction: String?,
            stationSequence: List<Station>
        ): Route = Route(
            origin = origin,
            destination = destination,
            directLine = line,
            transferLines = emptyList(),
            requiresTransfer = false,
            direction = direction,
            lines = listOf(line),
            transferStations = emptyList(),
            stationSequencesByLine = mapOf(line to stationSequence)
        )

        @JvmStatic
        fun transfer(
            origin: Station,
            destination: Station,
            lines: List<Line>,
            transferStations: List<Station>,
            direction: String?,
            stationSequencesByLine: Map<Line, List<Station>>
        ): Route {
            require(lines.isNotEmpty()) { "A transfer route needs at least one line" }
            require(transferStations.size == lines.size - 1) {
                "A transfer route needs one transfer station between each line"
            }
            return Route(
                origin = origin,
                destination = destination,
                directLine = lines.first(),
                transferLines = lines.drop(1),
                requiresTransfer = true,
                direction = direction,
                lines = lines,
                transferStations = transferStations,
                stationSequencesByLine = stationSequencesByLine
            )
        }
    }
}

private fun <T> immutableRouteList(values: Collection<T>): List<T> =
    Collections.unmodifiableList(ArrayList(values))

private fun <K, V> immutableRouteMap(values: Map<K, List<V>>): Map<K, List<V>> =
    Collections.unmodifiableMap(
        LinkedHashMap(values.mapValues { (_, value) -> immutableRouteList(value) })
    )
